builder.go

```diff
@@ -82,83 +88,24 @@ func (b *Builder) Build() (*BuildMetadata, error) {
 	var labels []Label
 
 	for _, bp := range b.Group.Group {
-		bpInfo, err := bp.Lookup(b.BuildpacksDir)
+		bpTOML, err := bp.Lookup(b.BuildpacksDir)
 		if err != nil {
 			return nil, err
 		}
-		bpDirName := launch.EscapeID(bp.ID)
-		bpLayersDir := filepath.Join(layersDir, bpDirName)
-		bpPlanDir := filepath.Join(planDir, bpDirName)
-		if err := os.MkdirAll(bpLayersDir, 0777); err != nil {
-			return nil, err
-		}
-
-		if err := os.MkdirAll(bpPlanDir, 0777); err != nil {
-			return nil, err
-		}
-		bpPlanPath := filepath.Join(bpPlanDir, "plan.toml")
-
-		foundPlan := plan.find(bp.noAPI().noHomepage())
-		if api.MustParse(bp.API).Equal(api.MustParse("0.2")) {
-			for i := range foundPlan.Entries {
-				foundPlan.Entries[i].convertMetadataToVersion()
-			}
-		}
-		if err := WriteTOML(bpPlanPath, foundPlan); err != nil {
-			return nil, err
-		}
 
-		cmd := exec.Command(
-			filepath.Join(bpInfo.Path, "bin", "build"),
-			bpLayersDir,
-			platformDir,
-			bpPlanPath,
-		)
-		cmd.Dir = appDir
-		cmd.Stdout = b.Out
-		cmd.Stderr = b.Err
-
-		if bpInfo.Buildpack.ClearEnv {
-			cmd.Env = b.Env.List()
-		} else {
-			cmd.Env, err = b.Env.WithPlatform(platformDir)
-			if err != nil {
-				return nil, err
-			}
-		}
-		cmd.Env = append(cmd.Env, EnvBuildpackDir+"="+bpInfo.Path)
-
-		if err := cmd.Run(); err != nil {
-			return nil, NewLifecycleError(err, ErrTypeBuildpack)
-		}
-		if err := setupEnv(b.Env, bpLayersDir); err != nil {
-			return nil, err
-		}
-		var bpPlanOut BuildpackPlan
-		if _, err := toml.DecodeFile(bpPlanPath, &bpPlanOut); err != nil {
+		br, err := bpTOML.Build(bp, plan, config)
+		if err != nil {
 			return nil, err
 		}
-		var bpBOM []BOMEntry
-		plan, bpBOM = plan.filter(bp, bpPlanOut)
-		bom = append(bom, bpBOM...)
 
-		var launch LaunchTOML
-		tomlPath := filepath.Join(bpLayersDir, "launch.toml")
-		if _, err := toml.DecodeFile(tomlPath, &launch); os.IsNotExist(err) {
-			continue
-		} else if err != nil {
-			return nil, err
-		}
-		for i := range launch.Processes {
-			launch.Processes[i].BuildpackID = bp.ID
-		}
-		procMap.add(launch.Processes)
-		slices = append(slices, launch.Slices...)
-		labels = append(labels, launch.Labels...)
+		bom = append(bom, br.BOM...)
+		labels = append(labels, br.Labels...)
+		plan = plan.filter(br.Met)
+		procMap.add(br.Processes)
+		slices = append(slices, br.Slices...)
 	}
 
-	if b.PlatformAPI.Compare(api.MustParse("0.4")) < 0 {
-		//plaformApiVersion is less than comparisonVersion
+	if b.PlatformAPI.Compare(api.MustParse("0.4")) < 0 { // PlatformAPI <= 0.3
```

Comment:
As I'm reading this, it seems like top level version must match metadata version, regardless of the platform API. Maybe it makes sense to validate this up front, so that the `convert` methods don't have to return an error.